### PR TITLE
[URL Connector] Better folders structure

### PR DESCRIPTION
**Pull request by lasryaric**
# Better folders structure for URL connectors

- Top level of the folders hierarchy is now the root of the webiste, even if the user provided URL isn't.
- Pages that have the same name as a folder lives inside the folder.

Screenshot 2024-01-12 at 16 45 28

### CODE REVIEW

File: connectors/src/connectors/webcrawler/temporal/activities.ts [hunk 1]
```diff
@@ -178,3 +204,20 @@ export function getFolderForUrl(url: string) {
...
   }
 }
+
+function isTopFolder(url: string) {
+  return new URL(url).pathname === "/";
+}
+
+// Normalizes a url path by removing trailing slashes and empty path parts (eg: //)
+function normalizeFolderUrl(url: string) {
+  const parsed = new URL(url);
```
Comment (spolu) [inline connectors/src/connectors/webcrawler/temporal/activities.ts:214]: Isn't there something on URL to achieve the same purpose?

File: connectors/src/connectors/webcrawler/temporal/activities.ts [hunk 2]
```diff
@@ -141,11 +141,37 @@ export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
...
+      connectorId: connector.id,
+      webcrawlerConfigurationId: webCrawlerConfig.id,
+    },
+  });
+  for (const page of pages) {
+    const folder = await WebCrawlerFolder.findOne({
+      where: {
+        connectorId: connector.id,
+        webcrawlerConfigurationId: webCrawlerConfig.id,
+        url: normalizeFolderUrl(page.url),
```
Comment (spolu) [inline connectors/src/connectors/webcrawler/temporal/activities.ts:162]: If you're calling it here we would expect you to call the same normalization code at the moment you insert the folder?

File: connectors/src/connectors/webcrawler/temporal/activities.ts [hunk 3]
```diff
@@ -141,11 +141,37 @@ export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
   );
 
   await crawler.run([webCrawlerConfig.url]);
+  await crawler.teardown();
+
+  // Adjusting page parent urls.
```
Comment (spolu) [inline connectors/src/connectors/webcrawler/temporal/activities.ts:146]: I thiiiink we can do it incrementally:
- Inserting a new WebCrawlFolder: check for existing pages and update their parents to it.
- Inserting a page, check for existing WebCrawlFolder parent and update.

**Review by spolu — COMMENTED** [review]
I think normalization/sanitization should happen at insertion (if you use it at query)? I presume URL has a way to do it? 

I believe the parent rewriting can be done incrementally which would prevent things moving around at the end of the sync + would be easier to understand as the code would be close to creation of the WebCrawlerPage/Folder

**Comment by spolu** [comment]
If there is no buil-tin URL sanitization let's try to think about everything we want to sanitize and add it in that function but also call it at all WebCrawlerPage/Folder insertions

**Comment by lasryaric** [comment]
“Url sanitation “ is not really a thing, so I already researched it, in the
URL module and read a good bunch of the crawler documentation, if not
everything, and did not find anything about it.

So if we want a folder like structure based on the / separator, we just
have to split on “/“.
There are cases where we end up with empty value in the array depending on
the existence of trailing slashes, and technically a double // can exists
and it would mess up our UI, hence the .filter() part of it.

So on sanitation it’s pretty simple and I should do it ourselves.

I agree that seeing the folders moving around is not great, so we can just
put every page in it’s own folder, that should work pretty nicely.

**Comment by lasryaric** [comment]
So if we want to move a page only if it “has a folder”, then we need to do
it at the end once we know the full graph, which is what we are doing here.

So I guess you mean only insert in the DB once we have the full graph?

File: connectors/src/connectors/webcrawler/temporal/activities.ts [hunk 4]
```diff
@@ -67,68 +72,106 @@ export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
...
-            sync_type: "batch",
-          },
-        });
+          const normalizedPageUrl = normalizeFolderUrl(request.url);
+          if (!page2normalizedFolder.has(normalizedPageUrl)) {
+            page2normalizedFolder.set(normalizedPageUrl, new Set());
+          }
+          page2normalizedFolder.get(normalizedPageUrl)?.add(request.url);
+
+          await upsertToDatasource({
```
Comment (spolu) [inline connectors/src/connectors/webcrawler/temporal/activities.ts:150]: We likely don't want to serialize calls to upsertToDatasource ?

Should we instead put an advisory lock around the Page/Folder insert? Not massively great either. Maybe having another queue to which you add the final upsert is the way to go?

**Review by spolu — COMMENTED** [review]
upsertToDataSource does not need to be serialized

File: connectors/src/connectors/webcrawler/temporal/activities.ts (hunk 4 above)
Comment (lasryaric) [reply connectors/src/connectors/webcrawler/temporal/activities.ts:150]: @spolu just realizing that we are updating pages parentURL which means they need to re-upserted or need to wait for the end before upserting any of them.

All of that to have the page in its folder. Is worth it?

Comment (spolu) [reply connectors/src/connectors/webcrawler/temporal/activities.ts:150]: Parents can be updated separately if needed :+1:

**Comment by lasryaric** [comment]
@spolu this is a good example where the URL is more interesting than the title.
There can be cases both ways, but I tried using titles again locally, and I am still more convinced by URL.
We can try title in production or you can come take a look on my screen and decide to revert if needed.

Screenshot 2024-01-16 at 11 01 44
Screenshot 2024-01-16 at 11 02 15

**Comment by spolu** [comment]
I'm onboard on using the URL :+1:
(note: in that screenhot we see https://sequelize.org/ under docs which looks like a bug)

Let's call the page /folder/folder/page if page never has a child and otherwise /folder/folder/page/_index  and move on with our lives? 

Please make sure to perfectly clearly dissociate our parent/page internalids and names and the original URL which should at all time remain untouched of course.

File: connectors/src/connectors/webcrawler/index.ts [hunk 5]
```diff
@@ -152,7 +152,7 @@ export async function retrieveWebcrawlerConnectorPermissions({
               new URL(page.url).pathname
                 .split("/")
                 .filter((x) => x)
-                .pop() || page.url,
```
Comment (lasryaric) [inline connectors/src/connectors/webcrawler/index.ts:155]: @spolu this was why the full URL was showing sometimes.

File: connectors/src/connectors/webcrawler/temporal/activities.ts [hunk 6]
```diff
@@ -39,6 +44,7 @@ export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
   let pageCount = 0;
   let errorCount = 0;
   const createdFolders = new Set<string>();
+  const processQueue = new PQueue({ concurrency: 10 });
```
Comment (spolu) [inline connectors/src/connectors/webcrawler/temporal/activities.ts]: Maybe a tiny bit less? Maybe 4 to begin with?

**Review by spolu — COMMENTED** [review]
As per IRL we're still not clear on the invariants of page vs folder formation. Let's write this down somewhere, propose 2 or 3 scenarios if you believe > 1 are valid. And let's converge on that before updating code.

**Comment by lasryaric** [comment]
Implementing case # 4 discussed [here](https://github.com/dust-tt/dust/issues/3166#issuecomment-1895517703).

# Screenshots

Screenshot 2024-01-17 at 15 15 44
Screenshot 2024-01-17 at 15 16 01
Screenshot 2024-01-17 at 15 16 16
Screenshot 2024-01-17 at 15 16 27
Screenshot 2024-01-17 at 15 16 46

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**